File: parking spot booking/frontend/src/services/parkingService.js
import apiClient from "./apiClient";
import { unwrap, toTrimmed, toNumber, toLocalDateTime, toIsoDate } from "./apiHelpers";

const normalizePublicStats = (stats = {}) => ({
  totalSpots: stats.totalSpots ?? stats.total_spots ?? 0,
  activeDrivers: stats.activeDrivers ?? stats.active_drivers ?? 0,
  totalLenders: stats.totalLenders ?? stats.total_lenders ?? 0,
  totalCities: stats.totalCities ?? stats.total_cities ?? 0,
});

const toSpotPayload = (payload = {}) => {
  const normalized = {
    spot_title: toTrimmed(payload.spot_title),
    description: toTrimmed(payload.description),
    address_line: toTrimmed(payload.address_line),
    locality: toTrimmed(payload.locality),
    city: toTrimmed(payload.city),
    state: toTrimmed(payload.state),
    pincode: toTrimmed(payload.pincode),
    latitude: toNumber(payload.latitude),
    longitude: toNumber(payload.longitude),
    vehicle_type_allowed: Array.isArray(payload.vehicle_type_allowed)
      ? payload.vehicle_type_allowed
      : undefined,
    total_slots: 1,
    spot_type: toTrimmed(payload.spot_type),
    image_url: toTrimmed(payload.image_url),
    day_of_week: toTrimmed(payload.day_of_week),
    start_time: toTrimmed(payload.start_time),
    end_time: toTrimmed(payload.end_time),
    pricing_type: toTrimmed(payload.pricing_type),
    base_hourly_rate: toNumber(payload.base_hourly_rate),
    peak_hour_rate: toNumber(payload.peak_hour_rate),
    special_day_rate: toNumber(payload.special_day_rate),
    effective_from: toIsoDate(payload.effective_from),
    effective_to: toIsoDate(payload.effective_to),
  };

  if (payload.available !== undefined || payload.is_available !== undefined) {
    normalized.available = Boolean(
      payload.available !== undefined ? payload.available : payload.is_available
    );
  }

  return Object.fromEntries(Object.entries(normalized).filter(([, value]) => value !== undefined));
};

export const parkingService = {
  async getAllSpots() {
    const res = await apiClient.get("/parking/all");
    return unwrap(res) || [];
  },

  async getPublicStats() {
    const res = await apiClient.get("/parking/stats");
    return normalizePublicStats(unwrap(res) || {});
  },

  async getFeaturedSpots() {
    const res = await apiClient.get("/parking/featured");
    return unwrap(res) || [];
  },

  async searchSpots(filters = {}) {
    const params = {
      query: filters.query || undefined,
      vehicleType: filters.vehicleType || undefined,
      maxPrice: filters.maxPrice || undefined,
      spotType: filters.spotType || undefined,
      minRating: filters.minRating || undefined,
      maxDistance: filters.maxDistance || undefined,
      sort: filters.sort || "nearest",
      startTime: filters.timeRange?.start || undefined,
      endTime: filters.timeRange?.end || undefined,
      status: filters.status || undefined,
    };
    const res = await apiClient.get("/parking/search", { params });
    return unwrap(res) || [];
  },

  async getSpotById(spotId) {
    const res = await apiClient.get(`/parking/spots/${spotId}`);
    return unwrap(res);
  },

  async getSpotBookedWindows(spotId, date) {
    const res = await apiClient.get(`/parking/spots/${spotId}/booked-windows`, {
      params: { date: toIsoDate(date) },
    });
    return unwrap(res) || [];
  },

  async getAlternativeSpots(spotId, startTime, endTime) {
    const res = await apiClient.get(`/parking/spots/${spotId}/alternatives`, {
      params: {
        startTime: toLocalDateTime(startTime),
        endTime: toLocalDateTime(endTime),
      },
    });
    return unwrap(res) || [];
  },

  async getReviewsBySpot(spotId) {
    const res = await apiClient.get(`/reviews/spot/${spotId}`);
    return unwrap(res) || [];
  },

  async getLenderSpots() {
    const res = await apiClient.get("/lender/spots");
    return unwrap(res) || [];
  },

  async getLenderSpotById(spotId) {
    const res = await apiClient.get(`/lender/spots/${spotId}`);
    return unwrap(res);
  },

  async addSpot(_lenderId, payload) {
    const res = await apiClient.post("/lender/spots", toSpotPayload(payload));
    return unwrap(res);
  },

  async updateSpot(spotId, payload) {
    const res = await apiClient.put(`/lender/spots/${spotId}`, toSpotPayload(payload));
    return unwrap(res);
  },

  async deleteSpot(spotId) {
    const res = await apiClient.delete(`/lender/spots/${spotId}`);
    return unwrap(res);
  },

  async updateSpotAvailability(spotId, payload = {}) {
    const body = {
      day_of_week: toTrimmed(payload.day_of_week) || "all",
      start_time: toTrimmed(payload.start_time) || "06:00",
      end_time: toTrimmed(payload.end_time) || "23:00",
      effective_from: toIsoDate(payload.effective_from),
      effective_to: toIsoDate(payload.effective_to),
      available: Boolean(
        payload.available !== undefined ? payload.available : payload.is_available
      ),
    };
    const res = await apiClient.put(`/lender/spots/${spotId}/availability`, body);
    return unwrap(res);
  },
};

File: parking spot booking (final project)/backend/src/main/java/com/parkingfinder/backend/service/ParkingService.java
package com.parkingfinder.backend.service;

import com.parkingfinder.backend.dto.parking.SpotAvailabilityRequest;
import com.parkingfinder.backend.dto.parking.SpotBookedWindowDto;
import com.parkingfinder.backend.dto.parking.SpotCreateRequest;
import com.parkingfinder.backend.dto.parking.SpotImageRequest;
import com.parkingfinder.backend.dto.parking.PublicStatsResponse;
import com.parkingfinder.backend.dto.parking.SpotResponse;
import com.parkingfinder.backend.dto.parking.SpotUpdateRequest;
import java.time.LocalDateTime;
import java.time.LocalDate;
import java.util.List;

/**
 * Parking-spot lifecycle and discovery: CRUD, search, availability,
 * pricing lookup, booked-window queries, and alternative-spot suggestions.
 */
public interface ParkingService {

    /** All approved spots (public listing). */
    List<SpotResponse> getAllSpots();

    /** Platform-wide public counters (spots, drivers, lenders, cities). */
    PublicStatsResponse getPublicStats();

    /** Top-rated, approved spots highlighted on the home page. */
    List<SpotResponse> getFeaturedSpots();

    /**
     * Full-text + filter search across parking spots.
     *
     * @param query       free-text search (city, locality, address)
     * @param vehicleType vehicle type filter (car, bike, etc.)
     * @param maxPrice    maximum hourly price
     * @param spotType    indoor / outdoor / covered
     * @param minRating   minimum average rating
     * @param maxDistance  maximum distance in km
     * @param sort        sort order: nearest, lowest_price, highest_rated, newest
     * @param startTime   optional availability window start
     * @param endTime     optional availability window end
     * @param status      spot status filter
     */
    List<SpotResponse> searchSpots(
        String query,
        String vehicleType,
        String maxPrice,
        String spotType,
        String minRating,
        String maxDistance,
        String sort,
        String startTime,
        String endTime,
        String status
    );

    /** Full spot details by ID (public). */
    SpotResponse getSpotById(String spotId);

    /** Active booked time-windows for a spot on a given date (used for slot-picker UI). */
    List<SpotBookedWindowDto> getSpotBookedWindows(String spotId, LocalDate date);

    /** Nearby alternative spots when the requested spot/time is unavailable. */
    List<SpotResponse> findAlternativeSpots(String spotId, LocalDateTime startTime, LocalDateTime endTime);

    /** All spots belonging to the currently-authenticated lender. */
    List<SpotResponse> getMyLenderSpots();

    /** Single spot detail for the current lender. */
    SpotResponse getMyLenderSpotById(String spotId);

    /** Create a new parking spot (requires lender role). */
    SpotResponse createSpot(SpotCreateRequest request);

    /** Update an existing spot's details. */
    SpotResponse updateSpot(String spotId, SpotUpdateRequest request);

    /** Delete a spot (only if no bookings exist). */
    void deleteSpot(String spotId);

    /** Set or update the weekly availability schedule for a spot. */
    void updateSpotAvailability(String spotId, SpotAvailabilityRequest request);

    /** Attach a new image to a spot. */
    void addSpotImage(String spotId, SpotImageRequest request);
}
